docs/design-notes/raft-in-scylla.md

```diff
@@ -0,0 +1,207 @@
+
+## Establishing group 0 in a fresh cluster
+
+When a Scylla node starts for the first time, it must join
+group 0. If there is an existing cluster, joining can be done by
+sending a request to add the starting node to the group
+configuration.
+
+But what should the node do if there is no running cluster yet?
+
+A special persistent state machine is responsible for establishing
+an initial Raft configuration in a fresh cluster.
+We often refer to it as "discovery" state machine.
+
+Raft group 0 has an id (UUID) just like any other group. After a
+node boots, this id is persisted in `scylla_local` system table.
+If this id is present, the node can start a Raft instance for
+the group using the last saved state in `raft`, `raft_snapshots`
+and `raft_config` system tables, which are all retrievable by
+group id.
+If a persisted id is missing, it means the node is bootstrapping
+and haven't joined Raft yet.
+
+To find out an existing cluster, or possibly create a new one,
+an instance of 'discovery' state machine is created. This state
+machine stores its intermediate state in `discovery` system table.
+
+The machine is initialized with initial contact points: the seeds
+parameter of `scylla.yaml` file. Then the machine runs a discovery
+loop, during which it contacts all initial seeds, and, by
+induction, all nodes known to the contacted nodes. The nodes
+exchange the peer information until a transitive closure of all
+peers is built - i.e. there are no new nodes discovered in the
+last loop iteration and all previously discovered nodes have been
+contacted.
+
+In the process, the discovery state machine may find a node
+with an existing Raft group, in which case it sends a configuration
+change request to join. Otherwise the node with the lowest Raft id
+establishes a new group.
+
+As you can see from this description, Raft in Scylla supports
+speedy, concurrent bootstraps of multiple nodes. As soon as
+streaming is ready to support it, all Scylla topology changes
+can be switched to a concurrent algorithm.
+
+# Using Group 0 to perform schema changes
+
+The main goal of adding Raft to Scylla is providing linearizability
+of schema changes. The group 0 Raft log is used to share all schema
+changes information across all cluster nodes, and perform schema changes
+in the same order on every node. However, while Raft guarantees
+that if an operation succeeds, it's stored in the logs on the
+majority of nodes, it doesn't provide an API for strictly ordered
+schema changes out of the box, for two reasons:
+    - Raft only stores a log of the operations. The operations
+      themselves - the schema changes - are applied to the client
+      state machine once they are committed to the log. In order
+      to construct a new operation it's necessary to read the
+      latest state of the client state machine, and Raft itself
+      doesn't guarantee that the state of the local
+      client state  machine is the *latest*.  E.g. if a client
```

Comment:
Well, it's not that Raft cannot guarantee this "strict order of schema changes" out of the box. It can, you just have to choose the commands wisely. The problem we have is that we chose our commands "unwisely". I.e. our schema change commands are "blind writes" (mutations). We could have chosen them differently: the commands could be read-modify-write operations, so the read would happen during `apply`, and not outside Raft, before the command is constructed, like we do it now.